In [200]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from itertools import cycle, islice
import datetime as datetime
import glob
from matplotlib import pyplot
import plotly.graph_objects as go
import plotly.io as pio
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import log_loss, make_scorer
import category_encoders as ce
from sklearn.ensemble import RandomForestRegressor
import warnings

%matplotlib inline

plt.style.use('default')
sns.set(style="whitegrid")
pd.options.display.float_format = '{:20,.2f}'.format
warnings.filterwarnings('ignore')
plt.rcParams['figure.figsize'] = (15, 5)

from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error
import pandas as pd
import numpy as np
%matplotlib inline
import matplotlib.pyplot as plt
from sklearn import tree
from catboost import CatBoostRegressor

In [201]:
# for preprocessing the data
from sklearn.preprocessing import StandardScaler
# the model
from sklearn.ensemble import RandomForestClassifier
# for combining the preprocess with model training
from sklearn.pipeline import make_pipeline
# for optimizing the hyperparameters of the pipeline
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import f1_score

In [202]:
import urllib.request
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier

In [203]:
train_labels = pd.read_csv('train_labels.csv',low_memory=False, dtype= {
    'damage_grade':'uint8'
}).set_index('building_id')
train_values = pd.read_csv('train_values.csv',low_memory=False, dtype= {
    'geo_level_1_id':'category', 
    'geo_level_2_id':'category',
    'geo_level_3_id':'category', 
    'count_floors_pre_eq':'uint8',
    'age':'uint16',
    'area_percentage':'uint16', 
    'height_percentage':'uint16', 
    'land_surface_condition':'category', 
    'foundation_type':'category',
    'roof_type':'category',
    'ground_floor_type':'category',
    'other_floor_type':'category',
    'position':'category',
    'plan_configuration':'category', 
    'has_superstructure_adobe_mud':'uint8',
    'has_superstructure_mud_mortar_stone':'uint8',
    'has_superstructure_stone_flag':'uint8',
    'has_superstructure_cement_mortar_stone':'uint8', 
    'has_superstructure_mud_mortar_brick':'uint8', 
    'has_superstructure_cement_mortar_brick':'uint8', 
    'has_superstructure_timber':'uint8', 
    'has_superstructure_bamboo':'uint8',
    'has_superstructure_rc_non_engineered':'uint8',
    'has_superstructure_rc_engineered':'uint8',
    'has_superstructure_other':'uint8', 
    'legal_ownership_status':'category',
    'count_families':'uint16', 
    'has_secondary_use':'uint8', 
    'has_secondary_use_agriculture':'uint8', 
    'has_secondary_use_hotel':'uint8',
    'has_secondary_use_rental':'uint8',
    'has_secondary_use_institution':'uint8',
    'has_secondary_use_school':'uint8', 
    'has_secondary_use_industry':'uint8', 
    'has_secondary_use_health_post':'uint8', 
    'has_secondary_use_gov_office':'uint8', 
    'has_secondary_use_use_police':'uint8', 
    'has_secondary_use_other':'uint8',
}).set_index('building_id')

pd.options.display.float_format = '{:20,.2f}'.format

In [204]:
train = train_values.merge(train_labels, on='building_id')
train

geo_level_1_id geo_level_2_id geo_level_3_id  count_floors_pre_eq  \
building_id                                                                     
802906                   6            487          12198                    2   
28830                    8            900           2812                    2   
94947                   21            363           8973                    2   
590882                  22            418          10694                    2   
201944                  11            131           1488                    3   
...                    ...            ...            ...                  ...   
688636                  25           1335           1621                    1   
669485                  17            715           2060                    2   
602512                  17             51           8163                    3   
151409                  26             39           1851                    2   
747594                  21              9           9101                    3   

             age  area_percentage  height_percentage land_surface_condition  \
building_id                                                                   
802906        30                6                  5                      t   
28830         10                8                  7                      o   
94947         10                5                  5                      t   
590882        10                6                  5                      t   
201944        30                8                  9                      t   
...          ...              ...                ...                    ...   
688636        55                6                  3                      n   
669485         0                6                  5                      t   
602512        55                6                  7                      t   
151409        10               14                  6                      t   
747594        10                7                  6                      n   

            foundation_type roof_type  ... has_secondary_use_hotel  \
building_id                            ...                           
802906                    r         n  ...                       0   
28830                     r         n  ...                       0   
94947                     r         n  ...                       0   
590882                    r         n  ...                       0   
201944                    r         n  ...                       0   
...                     ...       ...  ...                     ...   
688636                    r         n  ...                       0   
669485                    r         n  ...                       0   
602512                    r         q  ...                       0   
151409                    r         x  ...                       0   
747594                    r         n  ...                       0   

            has_secondary_use_rental has_secondary_use_institution  \
building_id                                                          
802906                             0                             0   
28830                              0                             0   
94947                              0                             0   
590882                             0                             0   
201944                             0                             0   
...                              ...                           ...   
688636                             0                             0   
669485                             0                             0   
602512                             0                             0   
151409                             0                             0   
747594                             0                             0   

            has_secondary_use_school  has_secondary_use_industry  \
building_id                             

In [205]:
#geo_level_1_id_ohe = pd.get_dummies(train.geo_level_1_id, prefix='geo_level_1_id')
#train = train.merge(geo_level_1_id_ohe, on='building_id')
train = train.drop('geo_level_1_id', axis=1)
#train = train.drop('count_floors_pre_eq', axis=1)

In [206]:
#geo_level_2_id_ohe = pd.get_dummies(train.geo_level_2_id, prefix='geo_level_2_id')
#train = train.merge(geo_level_2_id_ohe, on='building_id')
train = train.drop('geo_level_2_id', axis=1)

In [207]:
#geo_level_3_id_ohe = pd.get_dummies(train.geo_level_3_id, prefix='geo_level_3_id')
#train = train.merge(geo_level_3_id_ohe, on='building_id')
train = train.drop('geo_level_3_id', axis=1)

In [208]:
#edad promedio segun cantidad de pisos
#train['edad_promedio'] = train.groupby('count_floors_pre_eq')['age'].transform('mean')

In [209]:
#cantidad de tipos de techo segun edad
#train['tipo_techo'] = train.groupby('age')['roof_type'].transform('count')
#train = train.drop('roof_type', axis=1)

In [210]:
#cantidad de familias por edificacion
#train['flias_afectadas'] = train.groupby('building_id')['count_families'].transform('count')

In [211]:
#train = train.drop('other_floor_type', axis=1)

In [212]:
other_floor_type_ohe = pd.get_dummies(train.other_floor_type, prefix='other_floor_type')
train = train.merge(other_floor_type_ohe, on='building_id')
train = train.drop('other_floor_type', axis=1)

In [213]:
position_ohe = pd.get_dummies(train.position, prefix='position')
train = train.merge(position_ohe, on='building_id')
train = train.drop('position', axis=1)

In [214]:
plan_configuration_ohe = pd.get_dummies(train.plan_configuration, prefix='plan_configuration')
train = train.merge(plan_configuration_ohe, on='building_id')
train = train.drop('plan_configuration', axis=1)

In [215]:
legal_ownership_status_ohe = pd.get_dummies(train.legal_ownership_status, prefix='legal_ownership_status')
train = train.merge(legal_ownership_status_ohe, on='building_id')
train = train.drop('legal_ownership_status', axis=1)

In [216]:
ground_floor_type_ohe = pd.get_dummies(train.ground_floor_type, prefix='ground_floor_type')
train = train.merge(ground_floor_type_ohe, on='building_id')
train = train.drop('ground_floor_type', axis=1)

In [217]:
foundation_type_ohe = pd.get_dummies(train.foundation_type, prefix='foundation_type')
train = train.merge(foundation_type_ohe, on='building_id')
train = train.drop('foundation_type', axis=1)

In [218]:
roof_type_ohe = pd.get_dummies(train.roof_type, prefix='roof_type')
train = train.merge(roof_type_ohe, on='building_id')
train = train.drop('roof_type', axis=1)

In [219]:
land_surface_condition_ohe = pd.get_dummies(train.land_surface_condition, prefix='land_surface_condition')
train = train.merge(land_surface_condition_ohe, on='building_id')
train = train.drop('land_surface_condition', axis=1)

In [220]:
train_value = train['damage_grade']
train_data = train.iloc[:, train.columns != 'damage_grade']

In [221]:
LogLoss = make_scorer(log_loss, greater_is_better=False, needs_proba=False)

In [222]:
from sklearn.model_selection import train_test_split
X_train, X_validation, y_train, y_validation = train_test_split(train_data, train_value, train_size=0.7, random_state=1234)

In [223]:
#categorical_features_indices = np.where(train_data.dtypes == np.object)[0]

In [224]:
modelo = RandomForestClassifier(random_state=11, n_estimators=75, min_samples_split = 5, min_samples_leaf = 1,
                               class_weight="balanced", max_features="auto", max_depth = 28, criterion = 'gini',
                               n_jobs = 4, verbose = 0)
#min_impurity_decrease': 0.0,
 #'min_impurity_split': None,
modelo.fit(X_train, y_train)
print(f"RF Accuracy: {modelo.score(X_train, y_train)}")
print(f"RF Accuracy: {modelo.score(X_validation, y_validation)}")

RF Accuracy: 0.7206885209955048
RF Accuracy: 0.5735920492191198


modelo = RandomForestClassifier(random_state=11, n_estimators=75, min_samples_split = 5, min_samples_leaf = 1,
                               class_weight="balanced", max_features="auto", max_depth = 28, criterion = 'gini',
                               n_jobs = 4, verbose = 0)
RF Accuracy: 0.7194660673171801
RF Accuracy: 0.6623220475563116                               

modelo = RandomForestClassifier(random_state=11, n_estimators=75, min_samples_split = 5, min_samples_leaf = 1,
                               class_weight="balanced", max_features="auto", max_depth = 28, criterion = 'gini',
                               n_jobs = 4, verbose = 0)
RF Accuracy: 0.7206885209955048
RF Accuracy: 0.5735920492191198

RandomForestClassifier(random_state=11, n_estimators=75, min_samples_split = 5, #min_samples_leaf = 1,
                               class_weight="balanced", max_features="auto", max_depth = 28, criterion = 'gini',
                               n_jobs = 4, verbose = 0)
RF Accuracy: 0.703892117092424
RF Accuracy: 0.5679385016819943

max_depth = 28
RF Accuracy: 0.6950772941563425
RF Accuracy: 0.5676059400621635

max_depth = 30
RF Accuracy: 0.7111117202061178
RF Accuracy: 0.5670303526432253

max_depth = 40
RF Accuracy: 0.7498793991886854
RF Accuracy: 0.5593686445555826

RF Accuracy: 0.7488817015678105
RF Accuracy: 0.5532162545887108

RF Accuracy: 0.6678215107992544
RF Accuracy: 0.5262787633824075

RandomForestClassifier(random_state=11, n_estimators=82, min_samples_split = 6, #min_samples_leaf = 1,
                               class_weight="balanced", max_features="auto", max_depth = 49, criterion = 'gini')
RF Accuracy: 0.6408946387457516
RF Accuracy: 0.5144856167099423
RandomForestClassifier(random_state=11, n_estimators=82, min_samples_split = 6, #min_samples_leaf = 1,
                               class_weight="balanced", max_features="auto", max_depth = 50, criterion = 'gini')
RF Accuracy: 0.6409987939918869
RF Accuracy: 0.5143065450684949

RandomForestClassifier(random_state=11, n_estimators=200, class_weight="balanced", max_features="log2") 
=> RF Accuracy: 0.5114669804684002
RandomForestClassifier(random_state=11, n_estimators=200, class_weight="balanced", max_features="log2",
                        min_samples_split = 2, min_samples_leaf = 1, oob_score = True)
=> RF Accuracy: 0.5114669804684002
RandomForestClassifier(random_state=11, n_estimators=100, min_samples_split = 10, min_samples_leaf = 1,
                               class_weight="balanced", max_features="log2", oob_score = True)
=> RF Accuracy: 0.5119402412350826
RandomForestClassifier(random_state=11, n_estimators=100, min_samples_split = 50, min_samples_leaf = 1,
                               class_weight="balanced", max_features="log2", oob_score = True)
=> RF Accuracy: 0.5079750834601757
RandomForestClassifier(random_state=11, n_estimators=100, min_samples_split = 20, min_samples_leaf = 1,
                               class_weight="balanced", max_features="log2", oob_score = True)
=> RF Accuracy: 0.5119274504035507
RandomForestClassifier(random_state=11, n_estimators=100, min_samples_split = 10, min_samples_leaf = 1,
                               class_weight="balanced", max_features="auto", oob_score = True)
=> RF Accuracy: 0.512784436116192
RandomForestClassifier(random_state=11, n_estimators=800, min_samples_split = 5, #min_samples_leaf = 1,
                               class_weight="balanced", max_features="auto", max_depth = 50)
=> RF Accuracy: 0.5128355994423197
RandomForestClassifier(random_state=11, n_estimators=80, min_samples_split = 5, #min_samples_leaf = 1,
                               class_weight="balanced", max_features="auto", max_depth = 50)
=> RF Accuracy: 0.5132193243882784
RandomForestClassifier(random_state=11, n_estimators=80, min_samples_split = 6, #min_samples_leaf = 1,
                               class_weight="balanced", max_features="auto", max_depth = 50)
=> RF Accuracy: 0.5140251467747918


In [225]:
test = pd.read_csv('test_values.csv',low_memory=False, dtype= {
    'geo_level_1_id':'category', 
    'geo_level_2_id':'category',
    'geo_level_3_id':'category', 
    'count_floors_pre_eq':'uint8',
    'age':'uint16',
    'area_percentage':'uint16', 
    'height_percentage':'uint16', 
    'land_surface_condition':'category', 
    'foundation_type':'category',
    'roof_type':'category',
    'ground_floor_type':'category',
    'other_floor_type':'category',
    'position':'category',
    'plan_configuration':'category', 
    'has_superstructure_adobe_mud':'uint8',
    'has_superstructure_mud_mortar_stone':'uint8',
    'has_superstructure_stone_flag':'uint8',
    'has_superstructure_cement_mortar_stone':'uint8', 
    'has_superstructure_mud_mortar_brick':'uint8', 
    'has_superstructure_cement_mortar_brick':'uint8', 
    'has_superstructure_timber':'uint8', 
    'has_superstructure_bamboo':'uint8',
    'has_superstructure_rc_non_engineered':'uint8',
    'has_superstructure_rc_engineered':'uint8',
    'has_superstructure_other':'uint8', 
    'legal_ownership_status':'category',
    'count_families':'uint16', 
    'has_secondary_use':'uint8', 
    'has_secondary_use_agriculture':'uint8', 
    'has_secondary_use_hotel':'uint8',
    'has_secondary_use_rental':'uint8',
    'has_secondary_use_institution':'uint8',
    'has_secondary_use_school':'uint8', 
    'has_secondary_use_industry':'uint8', 
    'has_secondary_use_health_post':'uint8', 
    'has_secondary_use_gov_office':'uint8', 
    'has_secondary_use_use_police':'uint8', 
    'has_secondary_use_other':'uint8',
}).set_index('building_id')

pd.options.display.float_format = '{:20,.2f}'.format

In [226]:
#geo_level_1_id_ohe = pd.get_dummies(test.geo_level_1_id, prefix='geo_level_1_id')
#test = test.merge(geo_level_1_id_ohe, on='building_id')
test = test.drop('geo_level_1_id', axis=1)

#geo_level_2_id_ohe = pd.get_dummies(test.geo_level_2_id, prefix='geo_level_2_id')
#test = test.merge(geo_level_2_id_ohe, on='building_id')
test = test.drop('geo_level_2_id', axis=1)

#geo_level_3_id_ohe = pd.get_dummies(test.geo_level_3_id, prefix='geo_level_3_id')
#test = test.merge(geo_level_3_id_ohe, on='building_id')
test = test.drop('geo_level_3_id', axis=1)

other_floor_type_ohe = pd.get_dummies(test.other_floor_type, prefix='other_floor_type')
test = test.merge(other_floor_type_ohe, on='building_id')
test = test.drop('other_floor_type', axis=1)

position_ohe = pd.get_dummies(test.position, prefix='position')
test = test.merge(position_ohe, on='building_id')
test = test.drop('position', axis=1)

plan_configuration_ohe = pd.get_dummies(test.plan_configuration, prefix='plan_configuration')
test = test.merge(plan_configuration_ohe, on='building_id')
test = test.drop('plan_configuration', axis=1)

legal_ownership_status_ohe = pd.get_dummies(test.legal_ownership_status, prefix='legal_ownership_status')
test = test.merge(legal_ownership_status_ohe, on='building_id')
test = test.drop('legal_ownership_status', axis=1)

ground_floor_type_ohe = pd.get_dummies(test.ground_floor_type, prefix='ground_floor_type')
test = test.merge(ground_floor_type_ohe, on='building_id')
test = test.drop('ground_floor_type', axis=1)

foundation_type_ohe = pd.get_dummies(test.foundation_type, prefix='foundation_type')
test = test.merge(foundation_type_ohe, on='building_id')
test = test.drop('foundation_type', axis=1)

roof_type_ohe = pd.get_dummies(test.roof_type, prefix='roof_type')
test = test.merge(roof_type_ohe, on='building_id')
test = test.drop('roof_type', axis=1)

land_surface_condition_ohe = pd.get_dummies(test.land_surface_condition, prefix='land_surface_condition')
test = test.merge(land_surface_condition_ohe, on='building_id')
test = test.drop('land_surface_condition', axis=1)

In [227]:
preds = modelo.predict(test)

In [230]:
preds

array([3, 2, 2, ..., 3, 2, 2], dtype=uint8)

In [231]:
ids = test.reset_index()['building_id']

In [234]:
ids

0         300051
1          99355
2         890251
3         745817
4         421793
          ...   
86863     310028
86864     663567
86865    1049160
86866     442785
86867     501372
Name: building_id, Length: 86868, dtype: int64

In [232]:
damages = pd.Series(preds).rename("damage_grade")

In [233]:
damages

0        3
1        2
2        2
3        1
4        3
        ..
86863    2
86864    3
86865    3
86866    2
86867    2
Name: damage_grade, Length: 86868, dtype: uint8

In [236]:
submit = pd.merge(ids, damages, left_index=True,right_index=True).set_index('building_id')

submit['damage_grade'] = submit['damage_grade'].round(0)
submit = submit.astype({'damage_grade':'int32'})
submit.to_csv('submit_randomforest_001.csv')